File: go/dsti-generator/generator.go
// Package generator generates Go test files from DSTI intent signals.
//
// @docspec:module {
//   id: "docspec-go-dsti-generator",
//   name: "DSTI Test Generator",
//   description: "Generates Go test stubs from DSTI intent graph JSON, producing guard clause tests and property-based tests (via rapid) derived from extracted intent signals.",
//   since: "3.0.0"
// }
package generator

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
)

// GeneratedTestFile represents a generated test file.
// @docspec:intentional "Captures the output of test generation: file path, content, and test category"
type GeneratedTestFile struct {
	Path     string
	Content  string
	TestType string
}

// GoTestGenerator generates Go test files.
//
// @docspec:boundary "DSTI-to-test-code generation"
// @docspec:intentional "Transforms DSTI intent signals into executable Go test stubs"
type GoTestGenerator struct {
	OutputDir string
}

// NewGoTestGenerator creates a new generator.
//
// @docspec:deterministic
func NewGoTestGenerator(outputDir string) *GoTestGenerator {
	return &GoTestGenerator{OutputDir: outputDir}
}

// Generate produces test files from intent graph JSON.
//
// @docspec:method { since: "3.0.0" }
// @docspec:intentional "Parses intent graph JSON and dispatches to guard test and property test generators based on signal presence"
func (g *GoTestGenerator) Generate(intentGraphJSON []byte) ([]GeneratedTestFile, error) {
	var graph struct {
		Methods []struct {
			Qualified     string                 `json:"qualified"`
			IntentSignals map[string]interface{} `json:"intentSignals"`
		} `json:"methods"`
	}

	if err := json.Unmarshal(intentGraphJSON, &graph); err != nil {
		return nil, fmt.Errorf("parse intent graph: %w", err)
	}

	var files []GeneratedTestFile

	for _, method := range graph.Methods {
		signals := method.IntentSignals
		qualified := method.Qualified

		// Guard tests
		if gc, ok := signals["guardClauses"]; ok {
			if count, ok := gc.(float64); ok && count > 0 {
				files = append(files, generateGuardTest(qualified, int(count), g.OutputDir))
			}
		}

		// Collect invariant rules
		var invariantRules []string
		if rules, ok := signals["invariantRules"].([]interface{}); ok {
			for _, r := range rules {
				if ruleStr, ok := r.(string); ok {
					invariantRules = append(invariantRules, ruleStr)
				}
			}
		}

		// Property tests
		if ns, ok := signals["nameSemantics"].(map[string]interface{}); ok {
			if intent, ok := ns["intent"].(string); ok {
				file := generatePropertyTest(qualified, intent, g.OutputDir, invariantRules)
				if file != nil {
					files = append(files, *file)
				}
			}
		} else if len(invariantRules) > 0 {
			file := generatePropertyTest(qualified, "", g.OutputDir, invariantRules)
			if file != nil {
				files = append(files, *file)
			}
		}
	}

	return files, nil
}

// @docspec:intentional "Generates guard clause test stubs from DSTI guardClauses signal count"
func generateGuardTest(qualified string, guardCount int, outputDir string) GeneratedTestFile {
	parts := strings.SplitN(qualified, "#", 2)
	typeName := "Unknown"
	methodName := "Unknown"
	if len(parts) >= 1 {
		typeParts := strings.Split(parts[0], ".")
		typeName = typeParts[len(typeParts)-1]
	}
	if len(parts) >= 2 {
		methodName = parts[1]
	}

	fileName := fmt.Sprintf("%s_%s_guard_test.go", strings.ToLower(typeName), strings.ToLower(methodName))

	var tests strings.Builder
	for i := 0; i < guardCount; i++ {
		tests.WriteString(fmt.Sprintf(`
func Test%s%s_Guard%d(t *testing.T) {
	// Given: input violating guard condition %d
	// When/Then: expect panic or error
	defer func() {
		if r := recover(); r == nil {
			t.Errorf("Expected panic for guard condition %d")
		}
	}()
	// TODO: Call %s with invalid input
}
`, typeName, methodName, i+1, i+1, i+1, methodName))
	}

	content := fmt.Sprintf(`package generated

import "testing"

// Guard clause tests for %s#%s.
// Auto-generated from DSTI intent signals.
%s`, typeName, methodName, tests.String())

	return GeneratedTestFile{
		Path:     filepath.Join(outputDir, fileName),
		Content:  content,
		TestType: "guard",
	}
}

// @docspec:intentional "Generates property-based test stubs using rapid from DSTI nameSemantics.intent and @DocInvariant rules"
func generatePropertyTest(qualified, intent, outputDir string, invariantRules []string) *GeneratedTestFile {
	parts := strings.SplitN(qualified, "#", 2)
	typeName := "Unknown"
	methodName := "Unknown"
	if len(parts) >= 1 {
		typeParts := strings.Split(parts[0], ".")
		typeName = typeParts[len(typeParts)-1]
	}
	if len(parts) >= 2 {
		methodName = parts[1]
	}

	fileName := fmt.Sprintf("%s_%s_property_test.go", strings.ToLower(typeName), strings.ToLower(methodName))

	var test string
	switch intent {
	case "query":
		test = fmt.Sprintf(`
func Test%s%s_Idempotent(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		input := rapid.String().Draw(t, "input")
		// result1 := sut.%s(input)
		// result2 := sut.%s(input)
		// assert result1 == result2
		_ = input
	})
}`, typeName, methodName, methodName, methodName)
	case "creation", "mutation":
		test = fmt.Sprintf(`
func Test%s%s_ChangesState(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		input := rapid.String().Draw(t, "input")
		// before := sut.GetState()
		// sut.%s(input)
		// after := sut.GetState()
		// assert before != after
		_ = input
	})
}`, typeName, methodName, methodName)
	}

	// Generate invariant rule tests from @DocInvariant Property DSL expressions
	for _, rule := range invariantRules {
		assertion := mapDslToGoAssertion(rule)
		safeName := sanitizeGoName(rule)
		test += fmt.Sprintf(`
func Test%s%s_Invariant_%s(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		input := rapid.String().Draw(t, "input")
		result := sut.%s(input)
		%s
		_ = result
	})
}`, typeName, methodName, safeName, methodName, assertion)
	}

	if test == "" {
		return nil
	}

	content := fmt.Sprintf(`package generated

import (
	"testing"
	"pgregory.net/rapid"
)

// Property-based tests for %s#%s.
// Auto-generated from DSTI intent signals using rapid.
%s
`, typeName, methodName, test)

	return &GeneratedTestFile{
		Path:     filepath.Join(outputDir, fileName),
		Content:  content,
		TestType: "property",
	}
}

// sanitizeGoName converts a DSL expression to a valid Go identifier fragment.
func sanitizeGoName(s string) string {
	re := regexp.MustCompile(`[^a-zA-Z0-9]+`)
	result := re.ReplaceAllString(s, "_")
	result = strings.Trim(result, "_")
	return result
}

// toGoAccessor converts a dotted field path to a Go accessor, mapping "output" to "result".
func toGoAccessor(field string) string {
	parts := strings.Split(field, ".")
	for i, part := range parts {
		if i == 0 && part == "output" {
			parts[i] = "result"
		} else if i > 0 && len(part) > 0 {
			parts[i] = strings.ToUpper(part[:1]) + part[1:]
		}
	}
	return strings.Join(parts, ".")
}

// mapDslToGoAssertion maps a Property DSL expression to a Go assertion string.
//
// Supports NOT_NULL, NOT_EMPTY, NOT_BLANK, SIZE, IN, BETWEEN, RANGE,
// MATCHES, comparison operators, and monotonicity patterns.
//
// @docspec:deterministic
func mapDslToGoAssertion(expression string) string {
	expr := strings.TrimSpace(expression)

	// RANGE shorthand: field RANGE min..max
	rangeRe := regexp.MustCompile(`(?i)^(.+)\s+RANGE\s+(-?\d+(?:\.\d+)?)\.\.(-?\d+(?:\.\d+)?)$`)
	if m := rangeRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(strings.TrimSpace(m[1]))
		return fmt.Sprintf("if %s < %s || %s > %s { t.Errorf(\"expected %s in range [%s, %s]\") }", accessor, m[2], accessor, m[3], accessor, m[2], m[3])
	}

	// Monotonicity: field UP/DOWN -> field UP/DOWN
	monoRe := regexp.MustCompile(`^(.+)\s+(UP|DOWN)\s*[→\->]+\s*(.+)\s+(UP|DOWN)$`)
	if m := monoRe.FindStringSubmatch(expr); m != nil {
		outputAccessor := toGoAccessor(strings.TrimSpace(m[3]))
		cmp := ">="
		if m[4] != "UP" {
			cmp = "<="
		}
		return fmt.Sprintf("baseline := %s\n\t\t// Increase %s and verify %s moves %s\n\t\tif !(%s %s baseline) { t.Errorf(\"monotonicity violated\") }", outputAccessor, strings.TrimSpace(m[1]), strings.TrimSpace(m[3]), m[4], outputAccessor, cmp)
	}

	// NOT_NULL
	notNullRe := regexp.MustCompile(`(?i)^(\S+)\s+NOT_NULL$`)
	if m := notNullRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		return fmt.Sprintf("if %s == nil { t.Fatal(\"expected non-nil\") }", accessor)
	}

	// NOT_EMPTY
	notEmptyRe := regexp.MustCompile(`(?i)^(\S+)\s+NOT_EMPTY$`)
	if m := notEmptyRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		return fmt.Sprintf("if len(%s) == 0 { t.Fatal(\"expected non-empty\") }", accessor)
	}

	// NOT_BLANK
	notBlankRe := regexp.MustCompile(`(?i)^(\S+)\s+NOT_BLANK$`)
	if m := notBlankRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		return fmt.Sprintf("if strings.TrimSpace(%s) == \"\" { t.Fatal(\"expected non-blank\") }", accessor)
	}

	// SIZE comparison
	sizeRe := regexp.MustCompile(`(?i)^(\S+)\s+SIZE\s*(>=?|<=?|==|!=)\s*(-?\d+(?:\.\d+)?)$`)
	if m := sizeRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		if m[2] == ">" && m[3] == "0" {
			return fmt.Sprintf("if len(%s) == 0 { t.Fatal(\"expected non-empty\") }", accessor)
		}
		return fmt.Sprintf("if !(len(%s) %s %s) { t.Fatalf(\"expected len %s %s\") }", accessor, m[2], m[3], m[2], m[3])
	}

	// IN [values]
	inRe := regexp.MustCompile(`(?i)^(\S+)\s+IN\s*\[(.+)]$`)
	if m := inRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		vals := strings.Split(m[2], ",")
		var conditions []string
		for _, v := range vals {
			trimmed := strings.Trim(strings.TrimSpace(v), "\"'")
			conditions = append(conditions, fmt.Sprintf("%s == \"%s\"", accessor, trimmed))
		}
		return fmt.Sprintf("if !(%s) { t.Fatal(\"expected value in set\") }", strings.Join(conditions, " || "))
	}

	// BETWEEN min AND max
	betweenRe := regexp.MustCompile(`(?i)^(\S+)\s+BETWEEN\s+(-?\d+(?:\.\d+)?)\s+AND\s+(-?\d+(?:\.\d+)?)$`)
	if m := betweenRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		return fmt.Sprintf("if %s < %s || %s > %s { t.Errorf(\"expected %s in range [%s, %s]\") }", accessor, m[2], accessor, m[3], accessor, m[2], m[3])
	}

	// MATCHES pattern
	matchesRe := regexp.MustCompile(`(?i)^(\S+)\s+MATCHES\s+(.+)$`)
	if m := matchesRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		pattern := strings.Trim(strings.TrimSpace(m[2]), "\"'")
		return fmt.Sprintf("if matched, _ := regexp.MatchString(`%s`, %s); !matched { t.Fatal(\"pattern mismatch\") }", pattern, accessor)
	}

	// Comparison: >=, <=, !=, ==, >, <
	compRe := regexp.MustCompile(`^(\S+)\s*(>=|<=|!=|==|>|<)\s*(.+)$`)
	if m := compRe.FindStringSubmatch(expr); m != nil {
		accessor := toGoAccessor(m[1])
		return fmt.Sprintf("if !(%s %s %s) { t.Fatalf(\"expected %s %s %s\") }", accessor, m[2], strings.TrimSpace(m[3]), accessor, m[2], strings.TrimSpace(m[3]))
	}

	return fmt.Sprintf("// Property: %s", expression)
}

File: packages/test-generator/csharp/Generator.cs
// @docspec:module {
//   id: "docspec-csharp-dsti-generator",
//   name: "DSTI Test Generator",
//   description: "Generates xUnit + Moq + FsCheck test stubs from DSTI intent graph signals, producing guard clause tests and property-based tests.",
//   since: "3.0.0"
// }

using System.Text.Json;
using System.Text.RegularExpressions;

namespace DocSpec.DstiGenerator;

/// <summary>Generate xUnit + Moq + FsCheck test stubs from DSTI signals.</summary>
// [DocIntentional("Transform DSTI intent graph JSON into executable xUnit/FsCheck test stubs")]
// [DocBoundary("code generation")]
public class CSharpTestGenerator
{
    private readonly string _outputDir;

    public CSharpTestGenerator(string outputDir = "Tests/Generated")
    {
        _outputDir = outputDir;
    }

    // [DocMethod(Since = "3.0.0")]
    // [DocIntentional("Parse intent graph JSON and produce guard tests and property-based tests for each method")]
    public List<GeneratedTestFile> Generate(JsonElement intentGraph)
    {
        var files = new List<GeneratedTestFile>();
        var methods = intentGraph.GetProperty("methods").EnumerateArray();

        foreach (var method in methods)
        {
            var qualified = method.GetProperty("qualified").GetString() ?? "unknown";
            var signals = method.GetProperty("intentSignals");

            if (signals.TryGetProperty("guardClauses", out var guards) && guards.GetInt32() > 0)
            {
                files.Add(GenerateGuardTest(qualified, guards.GetInt32()));
            }

            var intent = signals.TryGetProperty("nameSemantics", out var ns) && ns.TryGetProperty("intent", out var i)
                ? i.GetString() : "unknown";

            var invariantRules = new List<string>();
            if (signals.TryGetProperty("invariantRules", out var rulesElement))
            {
                foreach (var ruleEl in rulesElement.EnumerateArray())
                {
                    var ruleStr = ruleEl.GetString();
                    if (!string.IsNullOrEmpty(ruleStr))
                        invariantRules.Add(ruleStr);
                }
            }

            if (intent == "query" || intent == "creation" || intent == "mutation" || invariantRules.Count > 0)
            {
                files.Add(GeneratePropertyTest(qualified, intent!, invariantRules));
            }
        }

        return files;
    }

    // [DocDeterministic]
    private GeneratedTestFile GenerateGuardTest(string qualified, int guardCount)
    {
        var parts = qualified.Split('#');
        var className = parts[0].Split('.').Last();
        var methodName = parts.Length > 1 ? parts[1] : "Unknown";
        var testClass = $"{className}{methodName}GuardTests";

        var tests = "";
        for (int i = 0; i < guardCount; i++)
        {
            tests += $@"
    [Fact]
    public void ShouldEnforceGuard{i + 1}()
    {{
        // Given: input violating guard condition {i + 1}
        // When/Then: expect exception
        Assert.Throws<ArgumentException>(() =>
        {{
            // TODO: Call {methodName} with invalid input
        }});
    }}
";
        }

        var content = $@"using Xunit;

/// <summary>Guard clause tests for {className}#{methodName}.</summary>
public class {testClass}
{{
{tests}
}}
";
        return new GeneratedTestFile($"{_outputDir}/{testClass}.cs", content, "guard");
    }

    // [DocDeterministic]
    private GeneratedTestFile GeneratePropertyTest(string qualified, string intent, List<string> invariantRules)
    {
        var parts = qualified.Split('#');
        var className = parts[0].Split('.').Last();
        var methodName = parts.Length > 1 ? parts[1] : "Unknown";
        var testClass = $"{className}{methodName}PropertyTests";

        var test = intent switch
        {
            "query" => $@"
    [Property]
    public Property QueryIsIdempotent(NonEmptyString input)
    {{
        // var result1 = sut.{methodName}(input.Get);
        // var result2 = sut.{methodName}(input.Get);
        // return (result1 == result2).ToProperty();
        return true.ToProperty();
    }}",
            "creation" or "mutation" => $@"
    [Property]
    public Property MutationChangesState(NonEmptyString input)
    {{
        // var before = sut.GetState();
        // sut.{methodName}(input.Get);
        // var after = sut.GetState();
        // return (before != after).ToProperty();
        return true.ToProperty();
    }}",
            _ => ""
        };

        // Generate invariant rule tests from @DocInvariant Property DSL expressions
        foreach (var rule in invariantRules)
        {
            var assertion = MapDslToCSharpAssertion(rule);
            var safeName = Regex.Replace(rule, @"[^a-zA-Z0-9]", "_");
            safeName = Regex.Replace(safeName, @"_+", "_").Trim('_');
            test += $@"

    [Fact]
    public void Invariant_{safeName}()
    {{
        var result = sut.{methodName}(/* input */);
        {assertion}
    }}";
        }

        var content = $@"using Xunit;
using FsCheck;
using FsCheck.Xunit;

/// <summary>Property-based tests for {className}#{methodName}.</summary>
public class {testClass}
{{
{test}
}}
";
        return new GeneratedTestFile($"{_outputDir}/{testClass}.cs", content, "property");
    }

    /// <summary>Maps a Property DSL expression to a C# assertion string.</summary>
    private static string MapDslToCSharpAssertion(string expression)
    {
        var expr = expression.Trim();

        // RANGE shorthand: field RANGE min..max
        var rangeMatch = Regex.Match(expr, @"(?i)^(.+)\s+RANGE\s+(-?\d+(?:\.\d+)?)\.\.(-?\d+(?:\.\d+)?)$");
        if (rangeMatch.Success)
        {
            var accessor = ToCSharpAccessor(rangeMatch.Groups[1].Value.Trim());
            return $"Assert.InRange({accessor}, {rangeMatch.Groups[2].Value}, {rangeMatch.Groups[3].Value});";
        }

        // Monotonicity: field UP/DOWN -> field UP/DOWN
        var monoMatch = Regex.Match(expr, @"^(.+)\s+(UP|DOWN)\s*[→\->]+\s*(.+)\s+(UP|DOWN)$");
        if (monoMatch.Success)
        {
            var outputAccessor = ToCSharpAccessor(monoMatch.Groups[3].Value.Trim());
            var cmp = monoMatch.Groups[4].Value == "UP" ? ">=" : "<=";
            return $"var baseline = {outputAccessor};\n        // Increase {monoMatch.Groups[1].Value.Trim()} and verify {monoMatch.Groups[3].Value.Trim()} moves {monoMatch.Groups[4].Value}\n        Assert.True({outputAccessor} {cmp} baseline);";
        }

        // NOT_NULL
        var notNullMatch = Regex.Match(expr, @"(?i)^(\S+)\s+NOT_NULL$");
        if (notNullMatch.Success)
            return $"Assert.NotNull({ToCSharpAccessor(notNullMatch.Groups[1].Value)});";

        // NOT_EMPTY
        var notEmptyMatch = Regex.Match(expr, @"(?i)^(\S+)\s+NOT_EMPTY$");
        if (notEmptyMatch.Success)
            return $"Assert.NotEmpty({ToCSharpAccessor(notEmptyMatch.Groups[1].Value)});";

        // NOT_BLANK
        var notBlankMatch = Regex.Match(expr, @"(?i)^(\S+)\s+NOT_BLANK$");
        if (notBlankMatch.Success)
            return $"Assert.False(string.IsNullOrWhiteSpace({ToCSharpAccessor(notBlankMatch.Groups[1].Value)}));";

        // SIZE comparison
        var sizeMatch = Regex.Match(expr, @"(?i)^(\S+)\s+SIZE\s*(>=?|<=?|==|!=)\s*(-?\d+(?:\.\d+)?)$");
        if (sizeMatch.Success)
        {
            var accessor = ToCSharpAccessor(sizeMatch.Groups[1].Value);
            var op = sizeMatch.Groups[2].Value;
            var val = sizeMatch.Groups[3].Value;
            if (op == ">" && val == "0")
                return $"Assert.NotEmpty({accessor});";
            return $"Assert.True({accessor}.Count {op} {val});";
        }

        // IN [values]
        var inMatch = Regex.Match(expr, @"(?i)^(\S+)\s+IN\s*\[(.+)]$");
        if (inMatch.Success)
        {
            var accessor = ToCSharpAccessor(inMatch.Groups[1].Value);
            var vals = string.Join(", ", inMatch.Groups[2].Value.Split(',')
                .Select(v =>
                {
                    var trimmed = v.Trim().Trim('"', '\'');
                    return int.TryParse(trimmed, out _) ? trimmed : $"\"{trimmed}\"";
                }));
            return $"Assert.Contains({accessor}, new[] {{ {vals} }});";
        }

        // BETWEEN min AND max
        var betweenMatch = Regex.Match(expr, @"(?i)^(\S+)\s+BETWEEN\s+(-?\d+(?:\.\d+)?)\s+AND\s+(-?\d+(?:\.\d+)?)$");
        if (betweenMatch.Success)
        {
            var accessor = ToCSharpAccessor(betweenMatch.Groups[1].Value);
            return $"Assert.InRange({accessor}, {betweenMatch.Groups[2].Value}, {betweenMatch.Groups[3].Value});";
        }

        // MATCHES pattern
        var matchesMatch = Regex.Match(expr, @"(?i)^(\S+)\s+MATCHES\s+(.+)$");
        if (matchesMatch.Success)
        {
            var accessor = ToCSharpAccessor(matchesMatch.Groups[1].Value);
            var pattern = matchesMatch.Groups[2].Value.Trim().Trim('"', '\'');
            return $"Assert.Matches(@\"{pattern}\", {accessor});";
        }

        // Comparison: >=, <=, !=, ==, >, <
        var compMatch = Regex.Match(expr, @"^(\S+)\s*(>=|<=|!=|==|>|<)\s*(.+)$");
        if (compMatch.Success)
        {
            var accessor = ToCSharpAccessor(compMatch.Groups[1].Value);
            var op = compMatch.Groups[2].Value;
            var val = compMatch.Groups[3].Value.Trim();
            if (op == "==")
            {
                var wrappedVal = int.TryParse(val, out _) ? val : $"\"{val}\"";
                return $"Assert.Equal({wrappedVal}, {accessor});";
            }
            if (op == "!=")
            {
                var wrappedVal = int.TryParse(val, out _) ? val : $"\"{val}\"";
                return $"Assert.NotEqual({wrappedVal}, {accessor});";
            }
            return $"Assert.True({accessor} {op} {val});";
        }

        return $"// Property: {expression}";
    }

    /// <summary>Converts a dotted field path to a C# accessor, mapping "output" to "result" and capitalizing properties.</summary>
    private static string ToCSharpAccessor(string field)
    {
        var parts = field.Split('.');
        var result = new List<string>();
        for (int i = 0; i < parts.Length; i++)
        {
            if (i == 0)
            {
                result.Add(parts[i] == "output" ? "result" : parts[i]);
            }
            else
            {
                result.Add(char.ToUpper(parts[i][0]) + parts[i][1..]);
            }
        }
        return string.Join(".", result);
    }
}

public record GeneratedTestFile(string Path, string Content, string Type);
